content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,142 @@
+searching for a local minimum.
+
+It is usually optimal to set $m_1$ and $m_2$ equidistant to each other and to
+the endpoints, so we can maximize the amount of distance we eliminate per
+search. We can set $m_1 = l + \frac{r-l}{3}$ and $m_2 = r - \frac{r-l}{3}$ so
+the intervals $[l, m_1], [m_1, m_2],$ and $[m_2,r]$ have roughly equal length.
+
+# Sample Problem - Haybale Distribution
+
+<FocusProblem problem="sample" />
+
+## Solution - Haybale Distribution
+
+Consider the function $f(y)$, which returns the number of haybales wasted if $y$
+is the chosen integer point. The problem asks for the local minimum for $f(y)$
+in the range of integers $[0, 10^6]$. It can be shown the function is unimodal
+because:
+
+- If we assume $y'$ is the optimal value of $y$ (i.e. the local minimum), it
+  holds that $f(x) \leq f(x+1)$ for all $x \in [0, y'-1]$. Similarily,
+  $f(x) \geq f(x-1)$ for all $x \in [y'+1, 10^6]$. In simpler terms, as you
+  slide $x$ away from $y'$, you are pulling away from the optimal value.
+
+Therefore, we can run a ternary search to find the local minimum. If
+$f(m_1) < f(m_2)$, then we can move the search range to $[l, m_2]$, otherwise,
+our search range is $[m_1, r]$.
+
+All that's left is computing $f(y)$. Let's sort the haybales first. We can
+obtain the greatest $i$ such that $x_i < f(y)$ using binary search. Then, we can
+obtain the sum of the piecewise function for haybales in the range $[1, i]$ and
+$[i+1, n]$ separately, using $a$ and $b$ respectively. This can be done with
+prefix sums. For more details, refer to the implementation.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+int main() {
+	int n;
+	cin >> n;
+	vector<ll> a(n);
+	for (int i = 0; i < n; i++) { cin >> a[i]; }
+	sort(a.begin(), a.end());
+
+	vector<ll> cum(n + 1);  // cumulative sums for a
+	for (int i = 1; i <= n; i++) { cum[i] = a[i - 1] + cum[i - 1]; }
+
+	int q;
+	cin >> q;
+	for (int query = 0; query < q; query++) {
+		ll left, right;
+		cin >> left >> right;
+
+		auto calc_cost = [&](ll x) -> ll {
+			int low_idx = lower_bound(a.begin(), a.end(), x) - a.begin() - 1;
+			ll left_cost = (ll)(low_idx + 1) * x - cum[low_idx + 1];
+
+			int high_idx = low_idx + 1;
+			ll right_cost =
+			    (ll)(cum[n] - cum[high_idx]) - (ll)(n - high_idx) * x;
+
+			return left * left_cost + right * right_cost;
+		};
+
+		ll l = 0, r = 1e6;
+		while (r - l > 3) {
+			ll mid1 = l + (r - l) / 3;
+			ll mid2 = r - (r - l) / 3;
+			if (calc_cost(mid1) > calc_cost(mid2)) {
+				l = mid1;
+			} else {
+				r = mid2;
+			}
+		}
+
+		ll ans = 1e18;
+		for (int i = l; i <= r; i++) { ans = min(ans, calc_cost(i)); }
+		cout << ans << "\n";
+	}
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+<Problems problems="probs" />
+
+Extra: Check out [Nested Ternary Search](https://en.wikipedia.org/wiki/Golden-section_search) and [Multidimensional Ternary Search](https://codeforces.com/blog/entry/98524).
```

Comment:
what is nested ternary search? golden-section search sounds like something different

Reply:
lol nvm its the same as multidimensional ternary search